File: backend/src/main/java/org/luxons/sevenwonders/game/moves/BuildWonderMove.java
package org.luxons.sevenwonders.game.moves;

import java.util.List;

import org.luxons.sevenwonders.game.Settings;
import org.luxons.sevenwonders.game.api.PlayerMove;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.cards.Card;

public class BuildWonderMove extends CardFromHandMove {

    BuildWonderMove(int playerIndex, Card card, PlayerMove move) {
        super(playerIndex, card, move);
    }

    @Override
    public boolean isValid(Table table, List<Card> playerHand) {
        if (!super.isValid(table, playerHand)) {
            return false;
        }
        Board board = table.getBoard(getPlayerIndex());
        return board.getWonder().isNextStageBuildable(table, getPlayerIndex(), getBoughtResources());
    }

    @Override
    public void place(Table table, List<Card> discardedCards, Settings settings) {
        Board board = table.getBoard(getPlayerIndex());
        board.getWonder().buildLevel(getCard().getBack());
    }

    @Override
    public void activate(Table table, List<Card> discardedCards, Settings settings) {
        int playerIndex = getPlayerIndex();
        Board board = table.getBoard(playerIndex);
        board.getWonder().activateLastBuiltStage(table, playerIndex, getBoughtResources());
    }
}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/engine/moves/BuildWonderMove.kt
package org.luxons.sevenwonders.engine.moves

import org.luxons.sevenwonders.engine.PlayerContext
import org.luxons.sevenwonders.engine.cards.Card
import org.luxons.sevenwonders.model.PlayerMove
import org.luxons.sevenwonders.model.Settings

internal class BuildWonderMove(move: PlayerMove, card: Card, player: PlayerContext) :
    CardFromHandMove(move, card, player) {

    private val wonder = player.board.wonder

    init {
        if (!wonder.isNextStageBuildable(playerContext.board, transactions)) {
            throw InvalidMoveException(this, "all levels are already built, or the given resources are insufficient")
        }
    }

    override fun place(discardedCards: MutableList<Card>, settings: Settings) = wonder.placeCard(card.back)

    override fun activate(discardedCards: List<Card>, settings: Settings) =
        wonder.activateLastBuiltStage(playerContext, transactions)
}
